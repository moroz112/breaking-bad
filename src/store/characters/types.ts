import { RequestState } from '../types';

export interface ICharacter {
  id: number;
  name: string;
  birthday: string;
  occupation: string[];
  img: string;
  status: string;
  nickname: string;
  appearance: string[];
  portrayed: string;
  category: string[];
}

export interface ICharactersState {
  currentCharacter: ICharacter | null;
  characterByIdGetState: RequestState | null;
  characterByIdGetError: string | null;
}

export const CHARACTER_GET_ONE = 'CHARACTER_GET_ONE';
export const CHARACTER_GET_ONE_SUCCESS = 'CHARACTER_GET_ONE_SUCCESS';
export const CHARACTER_GET_ONE_ERROR = 'CHARACTER_GET_ONE_ERROR';

// Get One Episode
export interface CharacterGetOneAction {
  type: typeof CHARACTER_GET_ONE;
  payload: string;
}

export interface CharacterGetOneSuccessAction {
  type: typeof CHARACTER_GET_ONE_SUCCESS;
  payload: ICharacter;
}

export interface CharacterGetOneErrorAction {
  type: typeof CHARACTER_GET_ONE_ERROR;
  payload: any;
}

export type CharactersActionTypes =
  | CharacterGetOneAction
  | CharacterGetOneSuccessAction
  | CharacterGetOneErrorAction;
