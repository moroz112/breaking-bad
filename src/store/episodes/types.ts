import { RequestState } from '../types';

export enum EpisodeSeries {
  BreakingBad = 'Breaking Bad',
  BetterCallSaul = 'Better Call Saul',
}

export interface IEpisode {
  episode_id: number;
  title: string;
  season: number;
  episode: number;
  air_date: string;
  characters: string[];
  series: EpisodeSeries;
}

export interface IEpisodesState {
  episodes: IEpisode[];
  episodesGetState: RequestState | null;
  episodesGetError: string | null;

  currentEpisode: IEpisode | null;
  episodeByIdGetState: RequestState | null;
  episodeByIdGetError: string | null;
}

export const EPISODES_GET = 'EPISODES_GET';
export const EPISODES_GET_SUCCESS = 'EPISODES_GET_SUCCESS';
export const EPISODES_GET_ERROR = 'EPISODES_GET_ERROR';

export const EPISODE_GET_ONE = 'EPISODE_GET_ONE';
export const EPISODE_GET_ONE_SUCCESS = 'EPISODE_GET_ONE_SUCCESS';
export const EPISODE_GET_ONE_ERROR = 'EPISODE_GET_ONE_ERROR';

export interface EpisodesGetAction {
  type: typeof EPISODES_GET;
  payload: EpisodeSeries;
}

export interface EpisodesGetSuccessAction {
  type: typeof EPISODES_GET_SUCCESS;
  payload: IEpisode[];
}

export interface EpisodesGetErrorAction {
  type: typeof EPISODES_GET_ERROR;
  payload: any;
}

export interface EpisodesGetOneAction {
  type: typeof EPISODE_GET_ONE;
  payload: number;
}

export interface EpisodesGetOneSuccessAction {
  type: typeof EPISODE_GET_ONE_SUCCESS;
  payload: IEpisode;
}

export interface EpisodesGetOneErrorAction {
  type: typeof EPISODE_GET_ONE_ERROR;
  payload: any;
}

export type EpisodesActionTypes =
  | EpisodesGetAction
  | EpisodesGetSuccessAction
  | EpisodesGetErrorAction
  | EpisodesGetOneAction
  | EpisodesGetOneSuccessAction
  | EpisodesGetOneErrorAction;
