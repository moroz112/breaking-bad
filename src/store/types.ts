import { ICharactersState } from './characters/types';
import { IEpisodesState } from './episodes/types';

export enum RequestState {
  Loading,
  Saving,
  Success,
  Error,
}

export interface AppStore {
  episodes: IEpisodesState;
  characters: ICharactersState;
}
