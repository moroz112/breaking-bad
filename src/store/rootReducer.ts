import { combineReducers } from 'redux';
import EpisodesReducer from './episodes/reducer';
import CharactersReducer from './characters/reducer';

const rootReducer = combineReducers({
  episodes: EpisodesReducer,
  characters: CharactersReducer,
});

export default rootReducer;
