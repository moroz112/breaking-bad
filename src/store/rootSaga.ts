import { all, fork } from 'redux-saga/effects';

import episodes from './episodes/sagas';
import characters from './characters/sagas';

// this is main root saga
export default function* root() {
  yield all([
      fork(episodes),
      fork(characters)]
  );
}
