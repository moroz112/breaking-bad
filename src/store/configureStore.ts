import thunk from 'redux-thunk';
import { createStore, applyMiddleware, compose } from 'redux';
import { createLogger } from 'redux-logger';
import createSagaMiddleware from 'redux-saga';

import rootReducer from './rootReducer';
import rootSaga from './rootSaga';

const sagaMiddleware = createSagaMiddleware();

const configureStore = (preloadedState: any) => {
  const enhancers = compose(
    applyMiddleware(thunk, createLogger(), sagaMiddleware),
  );
  const store = createStore(rootReducer, preloadedState, enhancers);
  return store;
};

const store = configureStore({});
sagaMiddleware.run(rootSaga);

export default store;
