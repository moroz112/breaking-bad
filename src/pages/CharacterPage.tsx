import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link, useParams } from 'react-router-dom';
import { characterGetOne } from '../store/characters/actions';
import { AppStore, RequestState } from '../store/types';

const CharacterPage = () => {
  const { id } = useParams();
  const dispatch = useDispatch();
  const { currentCharacter, characterByIdGetState } = useSelector(
    (store: AppStore) => store.characters,
  );

  useEffect(() => {
    id && dispatch(characterGetOne(id));
  }, [dispatch, id]);

  if (characterByIdGetState === RequestState.Loading || !currentCharacter) {
    return <>Loading...</>;
  }

  return (
    <div className="max-w-xs mx-auto mt-16">
        <div className="photo-wrapper p-2">
          <img
            className="w-80 h-80 mx-auto object-cover"
            src={currentCharacter.img}
            alt={currentCharacter.nickname}
          />
        </div>
        <div className="p-2">
          <h3 className="text-center text-xl text-gray-900 font-medium leading-8">
            {currentCharacter.name}
          </h3>
          <div className="text-center text-gray-400 text-xs">
            <p>{currentCharacter.status}</p>
          </div>
          <table className="text-xs">
            <tbody>
              <tr>
                <td className="px-4 py-4 font-bold">Birth</td>
                <td className="px-4 py-4">{currentCharacter.birthday}</td>
              </tr>
              <tr>
                <td className="px-4 py-4 font-bold">Nickname</td>
                <td className="px-4 py-4">{currentCharacter.nickname}</td>
              </tr>
              <tr>
                <td className="px-4 py-4 font-bold">Occupations</td>
                <td className="px-4 py-4">{currentCharacter.occupation.join(', ')}</td>
              </tr>
            </tbody>
          </table>

          <div className="text-center">
            <Link
              to="/"
              className="text-xs text-indigo-500 hover:underline hover:text-indigo-600 font-medium"
            >
              Episodes character appears in
            </Link>
          </div>
        </div>
    </div>
  );
};

export default CharacterPage;
