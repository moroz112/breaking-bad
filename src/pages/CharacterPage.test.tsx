import CharacterPage from './CharacterPage';
import React from 'react';
import renderer from 'react-test-renderer';
import { useDispatch, useSelector } from 'react-redux';
import { RequestState } from '../store/types';
import { useParams } from 'react-router-dom';
jest.mock('react-redux');
jest.mock('react-router');

const mockData = {
  characterByIdGetState: RequestState.Success,
  currentCharacter: {
    char_id: 1,
    name: 'Walter White',
    birthday: '09-07-1958',
    occupation: ['High School Chemistry Teacher', 'Meth King Pin'],
    img: 'https://images.amcnetworks.com/amc.com/wp-content/uploads/2015/04/cast_bb_700x1000_walter-white-lg.jpg',
    status: 'Presumed dead',
    nickname: 'Heisenberg',
    appearance: [1, 2, 3, 4, 5],
    portrayed: 'Bryan Cranston',
    category: 'Breaking Bad',
    better_call_saul_appearance: [],
  },
};
beforeEach(() => {
  (useParams as any).mockImplementation(() => ({ id: '1' }));
  (useDispatch as any).mockImplementation(() => () => {});
});

test('CharacterPage must render loading', () => {
  (useSelector as any).mockImplementation(() => ({ characterByIdGetState: RequestState.Loading }));
  const component = renderer.create(<CharacterPage />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('CharacterPage must render data', async () => {
  (useSelector as any).mockImplementation(() => mockData);
  const component = renderer.create(<CharacterPage />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
