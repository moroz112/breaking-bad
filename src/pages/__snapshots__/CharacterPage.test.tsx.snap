// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterPage must render data 1`] = `
<div
  className="max-w-xs mx-auto mt-16"
>
  <div
    className="photo-wrapper p-2"
  >
    <img
      alt="Heisenberg"
      className="w-80 h-80 mx-auto object-cover"
      src="https://images.amcnetworks.com/amc.com/wp-content/uploads/2015/04/cast_bb_700x1000_walter-white-lg.jpg"
    />
  </div>
  <div
    className="p-2"
  >
    <h3
      className="text-center text-xl text-gray-900 font-medium leading-8"
    >
      Walter White
    </h3>
    <div
      className="text-center text-gray-400 text-xs"
    >
      <p>
        Presumed dead
      </p>
    </div>
    <table
      className="text-xs"
    >
      <tbody>
        <tr>
          <td
            className="px-4 py-4 font-bold"
          >
            Birth
          </td>
          <td
            className="px-4 py-4"
          >
            09-07-1958
          </td>
        </tr>
        <tr>
          <td
            className="px-4 py-4 font-bold"
          >
            Nickname
          </td>
          <td
            className="px-4 py-4"
          >
            Heisenberg
          </td>
        </tr>
        <tr>
          <td
            className="px-4 py-4 font-bold"
          >
            Occupations
          </td>
          <td
            className="px-4 py-4"
          >
            High School Chemistry Teacher, Meth King Pin
          </td>
        </tr>
      </tbody>
    </table>
    <div
      className="text-center"
    >
      <a
        className="text-xs text-indigo-500 hover:underline hover:text-indigo-600 font-medium"
        onClick={[Function]}
      >
        Episodes character appears in
      </a>
    </div>
  </div>
</div>
`;

exports[`CharacterPage must render loading 1`] = `"Loading..."`;
